Cannot compile with ghc 9.2 because of use of Option datatype

Seems still to be using deprecated (now removed) Option.

```
[ 2 of 11] Compiling Data.Monoid.DecidablyEmpty ( src/Data/Monoid/DecidablyEmpty.hs, dist/build/Data/Monoid/DecidablyEmpty.o, dist/build/Data/Monoid/DecidablyEmpty.dyn_o )

src/Data/Monoid/DecidablyEmpty.hs:72:41: error:
    Not in scope: type constructor or class ‘Option’
   |
72 | instance Semigroup a => DecidablyEmpty (Option a) where
   |                                         ^^^^^^

src/Data/Monoid/DecidablyEmpty.hs:73:12: error:
    Not in scope: data constructor ‘Option’
   |
73 |   isEmpty (Option a) = isNothing a
   |            ^^^^
```

Other projects have used CPP for this - don't know if preference is to remove or use CPP.

--- a/.github/workflows/haskell.yml
+++ b/.github/workflows/haskell.yml
@@ -6,7 +6,7 @@ jobs:
   build:
     strategy:
       matrix:
-        ghc: ['8.0.2', '8.2.2', '8.4.4', '8.6.5', '8.8.4', '8.10.2', '9.0.1']
+        ghc: ['8.0.2', '8.2.2', '8.4.4', '8.6.5', '8.8.4', '8.10.7', '9.0.1', '9.2.2']
         os: ['ubuntu-latest', 'macos-latest']
         exclude:
           # There are some linker warnings in 802 on darwin that

--- a/ChangeLog.md
+++ b/ChangeLog.md
@@ -1,5 +1,9 @@
 # Revision history for patch
 
+## Unreleased
+
+* Support GHC 9.2
+
 ## 0.0.5.2 - 2022-01-09
 
 * Correct field order of `PatchMapWithMove.NodeInfo`.

--- a/patch.cabal
+++ b/patch.cabal
@@ -33,12 +33,12 @@ flag split-these
 library
   hs-source-dirs: src
   default-language: Haskell2010
-  build-depends: base >= 4.9 && < 4.16
+  build-depends: base >= 4.9 && < 4.17
                , constraints-extras >= 0.3 && < 0.4
                , containers >= 0.6 && < 0.7
                , dependent-map >= 0.3 && < 0.5
                , dependent-sum >= 0.6 && < 0.8
-               , lens >= 4.7 && < 5.1
+               , lens >= 4.7 && < 5.2
                , indexed-traversable >= 0.1 && < 0.2
                , semigroupoids >= 4.0 && < 6
                , transformers >= 0.5.6.0 && < 0.6

--- a/patch.cabal
+++ b/patch.cabal
@@ -22,7 +22,7 @@ extra-source-files:
   ChangeLog.md
 
 tested-with:
-  GHC ==8.0.2 || ==8.2.2 || ==8.4.4 || ==8.6.5 || ==8.8.1
+  GHC ==8.0.2 || ==8.2.2 || ==8.4.4 || ==8.6.5 || ==8.8.4 || ==8.10.7 || ==9.0.1 || ==9.2.2
   GHCJS ==8.4
 
 flag split-these

--- a/patch.cabal
+++ b/patch.cabal
@@ -41,7 +41,7 @@ library
                , lens >= 4.7 && < 5.2
                , indexed-traversable >= 0.1 && < 0.2
                , semigroupoids >= 4.0 && < 6
-               , transformers >= 0.5.6.0 && < 0.6
+               , transformers >= 0.5.6.0 && < 0.7
                , witherable >= 0.3 && < 0.5
 
   if impl(ghc < 8.6) -- really, if base < 8.12
@@ -92,7 +92,7 @@ test-suite hlint
                , directory
                , filepath
                , filemanip
-               , hlint (< 2.1 || >= 2.2.2) && < 3.4
+               , hlint (< 2.1 || >= 2.2.2) && < 3.5
   if impl(ghcjs)
     buildable: False
 

--- a/src/Data/Functor/Misc.hs
+++ b/src/Data/Functor/Misc.hs
@@ -50,6 +50,7 @@ import Data.GADT.Compare
 import Data.GADT.Show
 import Data.IntMap (IntMap)
 import qualified Data.IntMap as IntMap
+import Data.Kind (Type)
 import Data.Map (Map)
 import qualified Data.Map as Map
 import Data.Some (Some, mkSome)
@@ -63,7 +64,7 @@ import Data.Typeable hiding (Refl)
 
 -- | 'Const2' stores a value of a given type 'k' and ensures that a particular
 -- type 'v' is always given for the last type parameter
-data Const2 :: * -> x -> x -> * where
+data Const2 :: Type -> x -> x -> Type where
   Const2 :: k -> Const2 k v v
   deriving (Typeable)
 
@@ -130,7 +131,7 @@ weakenDMapWith f = Map.fromDistinctAscList . map (\(k :=> v) -> (mkSome k, f v))
 -- | 'WrapArg' can be used to tag a value in one functor with a type
 -- representing another functor.  This was primarily used with dependent-map <
 -- 0.2, in which the value type was not wrapped in a separate functor.
-data WrapArg :: (k -> *) -> (k -> *) -> * -> * where
+data WrapArg :: (k -> Type) -> (k -> Type) -> Type -> Type where
   WrapArg :: f a -> WrapArg g f (g a)
 
 deriving instance Eq (f a) => Eq (WrapArg g f (g' a))

--- a/src/Data/Monoid/DecidablyEmpty.hs
+++ b/src/Data/Monoid/DecidablyEmpty.hs
@@ -69,8 +69,10 @@ instance DecidablyEmpty (First a) where
 instance DecidablyEmpty (Last a) where
   isEmpty (Last a) = isNothing a
 deriving instance DecidablyEmpty a => DecidablyEmpty (Identity a)
+#if !MIN_VERSION_base(4,16,0)
 instance Semigroup a => DecidablyEmpty (Option a) where
   isEmpty (Option a) = isNothing a
+#endif
 deriving instance DecidablyEmpty m => DecidablyEmpty (WrappedMonoid m)
 instance (Ord a, Bounded a) => DecidablyEmpty (Max a)
 instance (Ord a, Bounded a) => DecidablyEmpty (Min a)

--- a/src/Data/Patch/DMapWithMove.hs
+++ b/src/Data/Patch/DMapWithMove.hs
@@ -33,6 +33,7 @@ import Data.Functor.Misc
 import Data.Functor.Product
 import Data.GADT.Compare (GEq (..), GCompare (..))
 import Data.GADT.Show (GShow, gshow)
+import Data.Kind (Type)
 import qualified Data.Map as Map
 import Data.Maybe
 import Data.Monoid.DecidablyEmpty
@@ -67,7 +68,7 @@ data NodeInfo k v a = NodeInfo
 
 -- |Structure describing a particular change to a key, be it inserting a new key (@From_Insert@), updating an existing key (@From_Insert@ again), deleting a
 -- key (@From_Delete@), or moving a key (@From_Move@).
-data From (k :: a -> *) (v :: a -> *) :: a -> * where
+data From (k :: a -> Type) (v :: a -> Type) :: a -> Type where
   -- |Insert a new or update an existing key with the given value @v a@
   From_Insert :: v a -> From k v a
   -- |Delete the existing key

